### PR TITLE
fix memory leaks in ks_asm().

### PR DESCRIPTION
**Pull request by quangnh89**

### CODE REVIEW

File: llvm/keystone/ks.cpp [hunk 1]
```diff
@@ -584,6 +612,10 @@ int ks_asm(ks_engine *ks,
     else {
         *insn_size = Msg.size();
         encoding = (unsigned char *)malloc(*insn_size);
+        if (!encoding) {
+            // TODO: set error
+            return -1;
```
Comment (aquynh) [inline llvm/keystone/ks.cpp]: at this place we can `return KS_ERR_NOMEM;`

File: llvm/keystone/ks.cpp [hunk 2]
```diff
@@ -555,7 +568,22 @@ int ks_asm(ks_engine *ks,
...
+        return -1;
+    }
     MCTargetAsmParser *TAP = ks->TheTarget->createMCAsmParser(*ks->STI, *Parser, *ks->MCII, ks->MCOptions);
+    if (!TAP) { 
+        // memory insufficient
+        delete Parser;
+        delete Streamer;
+        delete CE;
+        // TODO: set error
+        return -1;
```
Comment (aquynh) [inline llvm/keystone/ks.cpp]: at this place we can `return KS_ERR_NOMEM;`

File: llvm/keystone/ks.cpp [hunk 3]
```diff
@@ -555,7 +568,22 @@ int ks_asm(ks_engine *ks,
     ks->SrcMgr.AddNewSourceBuffer(std::move(*BufferPtr), SMLoc());
 
     MCAsmParser *Parser = createMCAsmParser(ks->SrcMgr, Ctx, *Streamer, *ks->MAI);
+    if (!Parser) {
+        delete Streamer;
+        delete CE;
+        // memory insufficient
+        // TODO: set error
+        return -1;
```
Comment (aquynh) [inline llvm/keystone/ks.cpp]: at this place we can `return KS_ERR_NOMEM;`

File: llvm/keystone/ks.cpp [hunk 4]
```diff
@@ -539,14 +539,27 @@ int ks_asm(ks_engine *ks,
...
+    }
     Streamer = ks->TheTarget->createMCObjectStreamer(
             Triple(ks->TripleName), Ctx, *ks->MAB, OS, CE, *ks->STI, ks->MCOptions.MCRelaxAll,
             /*DWARFMustBeAtTheEnd*/ false);
+            
+    if (!Streamer) {
+        // memory insufficient
+        delete CE;
+        // TODO: set error
+        return -1;
```
Comment (aquynh) [inline llvm/keystone/ks.cpp]: this can also return KS_ERR_NOMEM

File: llvm/keystone/ks.cpp [hunk 5]
```diff
@@ -539,14 +539,27 @@ int ks_asm(ks_engine *ks,
     MCContext Ctx(ks->MAI, ks->MRI, &ks->MOFI, &ks->SrcMgr, true, address);
     ks->MOFI.InitMCObjectFileInfo(Triple(ks->TripleName), Ctx);
     CE = ks->TheTarget->createMCCodeEmitter(*ks->MCII, *ks->MRI, Ctx);
-
+    if (!CE) {
+        // memory insufficient
+        // TODO: set error
+        return -1;
```
Comment (aquynh) [inline llvm/keystone/ks.cpp]: same here: we can return KS_ERR_NOMEM

**Comment by quangnh89** [comment]
Thank you. I will fix it.

**Comment by aquynh** [comment]
next time you dont need to close this PR. you can just make another fix in your local repo, commit it, then push to your repo, and that new commit will automatically show up here.

**Pull request closed without merging**